File: src/app/api/calendar/ics/route.ts
// src/app/api/calendar/ics/route.ts
import { NextRequest, NextResponse } from 'next/server';

export const dynamic = 'force-dynamic';

function nextDayISO(dateISO: string): string {
  const [y, m, d] = dateISO.split('-').map(Number);
  const dt = new Date(Date.UTC(y, m - 1, d + 1));
  const yyyy = dt.getUTCFullYear();
  const mm = String(dt.getUTCMonth() + 1).padStart(2, '0');
  const dd = String(dt.getUTCDate()).padStart(2, '0');
  return `${yyyy}-${mm}-${dd}`;
}

function toICSDate(iso: string): string {
  return iso.replace(/-/g, '');
}

export async function GET(req: NextRequest) {
  const { searchParams } = new URL(req.url);
  const title = searchParams.get('title') || 'Transit';
  const date = searchParams.get('date'); // YYYY-MM-DD
  const desc = searchParams.get('desc') || 'Astrology transit (wellness/entertainment).';

  if (!date || !/^\d{4}-\d{2}-\d{2}$/.test(date)) {
    return NextResponse.json({ ok: false, error: 'Missing or invalid date' }, { status: 400 });
  }

  const dtStart = toICSDate(date);
  const dtEnd = toICSDate(nextDayISO(date));

  const ics = [
    'BEGIN:VCALENDAR',
    'VERSION:2.0',
    'PRODID:-//NatalTransitsAI//EN',
    'CALSCALE:GREGORIAN',
    'METHOD:PUBLISH',
    'BEGIN:VEVENT',
    `UID:${Date.now()}@natal-transits`,
    `DTSTAMP:${toICSDate(new Date().toISOString().slice(0,10))}T000000Z`,
    `DTSTART;VALUE=DATE:${dtStart}`,
    `DTEND;VALUE=DATE:${dtEnd}`,
    `SUMMARY:${title.replace(/\r?\n/g, ' ')}`,
    `DESCRIPTION:${desc.replace(/\r?\n/g, ' ')}`,
    'END:VEVENT',
    'END:VCALENDAR',
  ].join('\r\n');

  return new NextResponse(ics, {
    status: 200,
    headers: {
      'Content-Type': 'text/calendar; charset=utf-8',
      'Content-Disposition': `attachment; filename="transit-${dtStart}.ics"`,
      'Cache-Control': 'no-store',
    },
  });
}

File: src/app/api/calendar/ics/route.cs
// src/app/api/calendar/ics/route.ts
export const runtime = 'nodejs';
export const dynamic = 'force-dynamic';

import { NextRequest } from 'next/server';

function pad(n: number, w = 2) {
  return String(n).padStart(w, '0');
}

function toICSStamp(d: Date) {
  // YYYYMMDDTHHMMSSZ (UTC)
  return (
    d.getUTCFullYear().toString() +
    pad(d.getUTCMonth() + 1) +
    pad(d.getUTCDate()) +
    'T' +
    pad(d.getUTCHours()) +
    pad(d.getUTCMinutes()) +
    pad(d.getUTCSeconds()) +
    'Z'
  );
}

export async function GET(req: NextRequest) {
  const { searchParams } = new URL(req.url);
  const title = searchParams.get('title') ?? 'Transit';
  const description = searchParams.get('description') ?? '';
  const startISO = searchParams.get('start'); // ISO es: 2025-09-18T09:00:00Z
  const durationMin = Number(searchParams.get('duration') ?? '60');

  if (!startISO) {
    return new Response('Missing start', { status: 400 });
  }

  const start = new Date(startISO);
  if (isNaN(start.getTime())) {
    return new Response('Invalid start', { status: 400 });
  }
  const end = new Date(start.getTime() + durationMin * 60_000);

  const now = new Date();
  const ics =
`BEGIN:VCALENDAR
VERSION:2.0
PRODID:-//NatalTransitsAI//EN
CALSCALE:GREGORIAN
METHOD:PUBLISH
BEGIN:VEVENT
UID:${start.getTime()}-nataltransitsai@local
DTSTAMP:${toICSStamp(now)}
DTSTART:${toICSStamp(start)}
DTEND:${toICSStamp(end)}
SUMMARY:${title.replace(/\r?\n/g, ' ')}
DESCRIPTION:${description.replace(/\r?\n/g, ' ')}
END:VEVENT
END:VCALENDAR
`;

  return new Response(ics, {
    status: 200,
    headers: {
      'Content-Type': 'text/calendar; charset=utf-8',
      'Content-Disposition': `attachment; filename="transit-${toICSStamp(start)}.ics"`,
    },
  });
}
